File: viz/spatial_viz.py
"""
Spatial audio visualiser — Python/pygame port of spatial-viz.js.

Draws a top-down view: listener head at centre, source dot at the
computed (azimuth, distance) position.

pygame.draw and pygame.font are injectable so tests can run without a
display by passing mock objects.
"""

import math


class SpatialViz:
    def __init__(self, surface, pygame_draw=None, pygame_font=None):
        """
        surface      : pygame.Surface (or MockSurface in tests).
        pygame_draw  : pygame.draw module (or mock).
        pygame_font  : pygame.font module (or mock).
        """
        self._surface = surface
        self._draw = pygame_draw
        self._font = pygame_font

    # ── Lazy import helpers ──────────────────────────────────────────────────

    @property
    def _pg_draw(self):
        if self._draw is None:
            import pygame
            self._draw = pygame.draw
        return self._draw

    @property
    def _pg_font(self):
        if self._font is None:
            import pygame
            self._font = pygame.font
        return self._font

    # ── Public API ───────────────────────────────────────────────────────────

    def update(self, azimuth: float, distance: float):
        width, height = self._surface.get_size()
        cx, cy = width / 2, height / 2
        max_dist = 10
        scale = (min(width, height) / 2 * 0.75) / max_dist

        # Clear
        self._surface.fill((0, 0, 0))

        # Source position
        rad = (azimuth * math.pi) / 180.0
        sx = cx + scale * distance * math.sin(rad)
        sy = cy - scale * distance * math.cos(rad)

        # Line: head → source
        self._pg_draw.line(
            self._surface, (85, 85, 85), (cx, cy), (sx, sy), 1
        )

        # Listener head circle
        self._pg_draw.circle(
            self._surface, (68, 170, 153), (int(cx), int(cy)), 12
        )

        # Source dot
        self._pg_draw.circle(
            self._surface, (238, 119, 68), (int(sx), int(sy)), 8
        )

        # Label
        font = self._pg_font.SysFont("monospace", 11)
        label = font.render(
            f"{round(azimuth)}\u00b0  {distance:.1f} m", True, (204, 204, 204)
        )
        self._surface.blit(label, (cx - label.get_width() / 2, height - 8))

    def clear(self):
        self._surface.fill((0, 0, 0))

File: client/spatial-viz.js
class SpatialViz {
  constructor(canvasEl) {
    this._canvas = canvasEl;
    this._ctx = canvasEl.getContext('2d');
  }

  update(azimuth, distance) {
    const { width, height } = this._canvas;
    const cx = width / 2;
    const cy = height / 2;
    const maxDist = 10;
    const scale = (Math.min(width, height) / 2 * 0.75) / maxDist;
    const ctx = this._ctx;

    ctx.clearRect(0, 0, width, height);

    // Source position
    const rad = (azimuth * Math.PI) / 180;
    const sx = cx + scale * distance * Math.sin(rad);
    const sy = cy - scale * distance * Math.cos(rad);

    // Arrow: head → source (line + arrowhead)
    const angle = Math.atan2(sy - cy, sx - cx);
    const headLen = 10;

    ctx.beginPath();
    ctx.moveTo(cx, cy);
    ctx.lineTo(sx, sy);
    ctx.strokeStyle = '#555';
    ctx.lineWidth = 1;
    ctx.stroke();

    ctx.beginPath();
    ctx.moveTo(sx, sy);
    ctx.lineTo(sx - headLen * Math.cos(angle - Math.PI / 6), sy - headLen * Math.sin(angle - Math.PI / 6));
    ctx.moveTo(sx, sy);
    ctx.lineTo(sx - headLen * Math.cos(angle + Math.PI / 6), sy - headLen * Math.sin(angle + Math.PI / 6));
    ctx.strokeStyle = '#888';
    ctx.lineWidth = 1.5;
    ctx.stroke();

    // Listener head
    ctx.beginPath();
    ctx.arc(cx, cy, 12, 0, Math.PI * 2);
    ctx.fillStyle = '#4a9';
    ctx.fill();

    // Source dot
    ctx.beginPath();
    ctx.arc(sx, sy, 8, 0, Math.PI * 2);
    ctx.fillStyle = '#e74';
    ctx.fill();

    // Label
    ctx.fillStyle = '#ccc';
    ctx.font = '11px monospace';
    ctx.textAlign = 'center';
    ctx.fillText(`${Math.round(azimuth)}°  ${distance.toFixed(1)} m`, cx, height - 8);
  }

  clear() {
    const { width, height } = this._canvas;
    this._ctx.clearRect(0, 0, width, height);
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { SpatialViz };
}
